In [ ]:
# установить свежую версию, чтобы использовать ф-ю wilcoxon
!pip install scipy==1.5.2 -q

     |████████████████████████████████| 25.9MB 1.4MB/s 
ERROR: tensorflow 2.2.0 has requirement scipy==1.4.1; python_version >= "3", but you'll have scipy 1.5.2 which is incompatible.
ERROR: albumentations 0.1.12 has requirement imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.9 which is incompatible.


In [ ]:
import math
import numpy as np
import pandas as pd
from scipy import stats as st
import matplotlib.pyplot as plt 

#### Введение

Дисперсионный анализ (ДА) (ANOVA) - это статистический метод используемый для выявления различий средних показателях нескольких групп.

Как правило ДА используется для измерения средних 3 и более групп.

**На практике:** Если мы хотим сравнить влияние нескольких лекарст на пациента, а может быть при этом учесть ещё пол и возраст. Либо хотим сравнить несколько методик обучения на успеваемость. Либо выявить влияет ли лаборатория в которой производились измерения на измерения.

В рамках ДА, температура тела, артериальное давление, успеваемость учеников или результаты анализов рассматриваются как **зависимые переменные**.

А лекарства, методика обучения, лаборатория, как **независимые переменные**, их так же называют факторами.

В случае, если мы исследуем зависимость от одной переменной, одного фактора, в этом случае будем применять методы однофакторного ДА.

Иначе будем использовать многофакторный ДА.

Различают так же многомерный и одномерный ДА, в зависимости от того, какой размерности у нас зависимая переменная.

Разработанные методы ДА, применимы так же для парных наблюдений.

В основе ДА лежит сравнение дисперсий внутригрупповых и межгрупповой. 

#### Модель



**Модель**

Предположим, что мы наблюдаем несколько зависимых переменных (температура, давление и т.д.) пусть их будет $i = \overline {1,m}$ на который влияет один фактор, который может принимать $k$ значений $A: A_1, ..., A_k$.

$x_{i, j} = m_i + a_{i,j} + \varepsilon_{i,j}$

$x_{i,j}$ наблюдения, $i$ номер параметра, $j$ фактор

$m_i$ истинное значение мат. ож. $i$-го параметра

$a_{i,j}$ влияние $j$ фактора на $i$ параметр

$\varepsilon_{i,j}$ ошибка наблюдений

#### Параметрический критерий. Однофакторный ДА

Будем предполагать, что зависимая переменная является одномерной. На значение зависимой переменной влияет только 1 фактор, принимающий $k\geq2$ значений.

На практике используют однофакторный ДА, когда $k \geq 3$.



Наблюдения соответствующие $j$ значению фактора, мы можем рассматривать как выборку из некоторой ген. совокупности. Таким образом мы будем рассматривать $k$ ген. совокупностей и $k$ выборок из этих ген. совокупностей.

Будем предполагать:



*   Выборки случайны и взаимно независимы
*   Генеральные совокупности имеют нормальный з.р.
*   Ген. сов. имеют равные дисперсии





$X_{11} \;\; X_{12} \;\; ... \;\; X_{1k}$

$X_{21} \;\; X_{22} \;\; ... \;\; X_{2k}$

$\;\; . \;\;\;\;\;\;\;\; . \;\;\;\; ... \;\;\;\; . \;\;$

$\;\; . \;\;\;\;\;\;\;\; . \;\;\;\; ... \;\;\;\; . \;\;$

$X_{n_11} \;\; X_{n_22} \;\; ... \;\; X_{n_kk}$

$N(a_1, \sigma) \;\; N(a_2, \sigma) \;\; ... \;\; N(a_k, \sigma)$

$\;\;\;\;\;\; \xi_1 \;\;\;\;\;\;\;\;\;\; \xi_2 \;\;\;\;\;\;\;\;\;\; ... \;\; \xi_k$

$x_{i, j}$ $i$-ый элемент, $j$-ой выборки


##### Гипотезы

Проверим гипотезу о равенстве мат. ож. генеральных совокупностей, если эта гипотеза выполняется, то все ген. совокупности имеют один и тот же закон распределения и следовательно фактор не влияет на зависимую переменную.

$H_0: a_1 = a_2 = ... = a_k$

$H_1: \exists t,s: a_t \neq a_s$

Будем рассматривать как каждую выборку по отдельности, так и одну большую объединенную выборку.

$N = \sum \limits_{j=1}^k n_j$ объём обобщённой выборки.

Внутригрупповое среднее (являются оценками $a_j$ мат. ож.)

$\overline X_j = \frac{1}{n_j} \sum \limits_{i=1}^{n_j} x_{i,j}$, где $j = \overline {1,k}$

Общее среднее

$\overline X = \frac{1}{N} \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} x_{i,j}$

Рассмотрим сумму квадратов отклонений всех наблюдений от общего среднего. Если выполняется гипотеза $H_0$, то разброс наблюдений относительно общего среднего $\overline X$ не будет велик.

$\sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X)^2$

для каждого $j$ добавим и вычтем внутригрупповое среднее $\overline X_j$, тогда

$\sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X)^2 = \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j}((x_{i,j} - \overline X_j) + (\overline X_j - \overline X))^2  $

возведем в квадрат

$\sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X)^2 = \\
\sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j}((x_{i,j} - \overline X_j) + (\overline X_j - \overline X))^2 = \\
 \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X_j)^2 + 2\sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X_j)(x_{i,j} - \overline X) +  \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (\overline X_j - \overline X)^2 = \\
 $

 оказывается, если провести некоторые преобразования, то центральное слагаемое будет равно 0.

 $\sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X_j)^2 +  \sum \limits_{j=1}^k n_j (\overline X_j - \overline X)^2$

 переобозначим

$Q_1 = \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X_j)^2$

$Q_2 = \sum \limits_{j=1}^k n_j (\overline X_j - \overline X)^2$

Если $H_0$ не выполняется, то значение $Q_1$ не будет менятся, т.к. данная величина характеризует разброс внутри группы (выборки).

Если $H_0$ не выполняется, то значение $Q_2$ увеличится. 

Независимо от того, выполняется $H_0$ или нет,  статистика

$\frac{Q_1}{N-k} \sim \sigma^2$ является точечной оценкой неизвестной дисперсии.

Статистика

$\frac{Q_2}{k-1} \sim \sigma^2$ будет являтся оценкой дисперсии, но только в случае, если выполняется $H_0$.

##### Статистика критерия $S_1$

Рассмотрим следующую сл. вел $\eta$ т.к. все ген. сов. имеют нормальный закон распределения, то величина будет иметь распределение

$\eta = \sum \limits_{i=1}^{n_j} \frac{ (x_{i,j} - \overline X_j)^2 } {\sigma^2} \sim \chi^2_{n_j - 1}$

Рассмотрим статистику $S_1$ т.к. предполагали, что все выборки взаимно независимы и имеют норм. закон распред, статистика будет иметь распределение

$S_1 = \frac{Q_1}{\sigma^2} = \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} \frac{ (x_{i,j} - \overline X_j)^2 } {\sigma^2} \sim \chi^2_{(N-k)}$

Рассмотрим сл. вел.

$\frac{\overline X_j - a_j}{\sigma} \sqrt{n_j} \sim N(0, 1)$

Распределения рассмотренных статистик не зависит от того, выполняется нулевая гипотеза или не выполняется.



##### Статистика критерия $S_2$

Пусть $H_0$ верна, рассмотрим статистику

$S_2 = \frac{Q_2}{\sigma^2} = \frac{1}{\sigma^2} \sum \limits_{j=1}^k n_j (\overline X_j - \overline X)^2 = \\
= \sum \limits_{j=1}^k \bigg (\frac{\sqrt{n_j}  (\overline X_j - a)}{\sigma} \bigg)^2 - \bigg (\sum \limits_{j=1}^k \sqrt{\frac{n_j}{N}} \frac{\sqrt{n_j} (\overline X_j - a)}{\sigma} \bigg)^2 \sim \chi^2_{k-1}$

##### Статистика критерия F

Статистики $S_1$ и $S_2$ взаимно независимы.

$F = \frac{S_2 / (k-1)}{S_1 / (N-k)} = \frac{Q_2 / (k-1)}{Q_1 / (N-k)} \sim F_{(k-1, N-k)}$

т.е. мы сравниваем две оценки неизвестной дисперсии $\sigma^2$. Если $H_0$ выполняется, то статистика $F$ имеет распределение Фишера. Если $H_0$ не выполняется, значения статистики $F$ возврастают.

Для проверки гипотезы о равенстве мат. ож. для заданного уровня значимости $\alpha$ мы будем строить критическую область как правостороннюю. 

##### Статистика критерия F в Scipy

В функцию передаем в качестве аргумента выборки st.f_oneway(group1, ..., groupk)

##### Критическая область

Критическая область для $\alpha$:

$V_k = (u_{1 - \alpha, k - 1, N - k}; +\infty)$, где

$u_{1 - \alpha, k - 1, N - k}$ - квантиль Фишера.

$F_в \in V_k \Rightarrow H_0$ отвергается

##### p-value

$p_{value} = 1 - F_{k-1, N-k}(F_в)$, где

$F_{k-1, N-k}$ распределение Фишера

$F_в$ выборочное значение статистики $F$

$p_{value} < \alpha \Rightarrow H_0$ отвергается

##### Задача №1

Зависит ли критическая частота мерцания от цвета радужной оболочки глах. Критической частотой мерцания называется такая частота мерцающего света, что при увеличении данной частоты глаз перестает различать мерцания и воспринимает мерцающий свет как непрерывный. 

Карие 26.8, 27.9, 23.7, 25, 26.3, 24.8, 25.7, 24.5

Зеленые 26.4, 26.2, 28, 26.9, 29.1

Голубое 25.7, 27.2, 29.9, 28.5, 29.4, 28.3

Зависит ли критическая частота мерцания от цвета радужной оболочки глаз?

**Решение:**

In [ ]:
X1 = np.array([26.8, 27.9, 23.7, 25, 26.3, 24.8, 25.7, 24.5])
X2 = np.array([26.4, 26.2, 28, 26.9, 29.1])
X3 = np.array([25.7, 27.2, 29.9, 28.5, 29.4, 28.3])
X = np.concatenate((X1, X2, X3), axis=0)
k = 3
n1, n2, n3 = len(X1), len(X2), len(X3)
N = n1 + n2 + n3
xm1, xm2, xm3, xm = X1.mean(), X2.mean(), X3.mean(), X.mean()

Вычислим статистики

$Q_1 = \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X_j)^2$

$Q_2 = \sum \limits_{j=1}^k n_j (\overline X_j - \overline X)^2$



In [ ]:
Q1 = sum((X1 - xm1)**2) + sum((X2 - xm2)**2) + sum((X3 - xm3)**2)
Q2 = n1 * (xm1 - xm)**2 + n2 * (xm2 - xm)**2 + n3 * (xm3 - xm)**2 

Найдём статистику $F$

$F = \frac{Q_2 / (k-1)}{Q_1 / (N-k)}$

In [ ]:
numerator = Q2 / (k-1)
denominator = Q1 / (N-k)
Fv = numerator / denominator
Fv

6.335270379037686

Зададим уровень значимости $\alpha = 0.05$. Найдём квантиль распределения Фишера для построения критической области.

$u_{1 - \alpha, k - 1, N - k}$

In [ ]:
alpha = 0.05
st.f.ppf(1 - alpha, k-1, N-k)

3.63372346759163

$V_k = (3.63, + \infty)$

$F_в \in V_k \Rightarrow H_0$ отвергается. Значит цвет радужной оболочки влияет на критическое значение мерцания, люди с разным цветом глаз по разному воспринимают мерцающий свет.

**Ответ:**

$H_0$ отвергается

#### Метод линейных контрастов

Если мы отклоняем $H_0$ при ДА, то мы скорее всего захотим выяснить мат. ож. каких именно генеральных совокупностей различается. В этом поможет данный метод.



**Линейный контраст $L_k$** - линейная комбинация мат. ож. $a_j$

$L_k = \sum \limits_{j=1}^k c_ja_j$, где

$c_j$ - коэффициенты, $\sum \limits_{j=1}^k c_j = 0$



т.к. мат. ож. нам не известно, тогда и линейный контраст нам не известен, давайте построим оценку точечную оценку линейного контраста

$\hat{L_k} = \sum \limits_{j=1}^k c_j \overline X_j$

т.к. выборки независимы, поэтому и внутригрупповые средние являются независимыми, поэтому $D(\xi_1 + ... + \xi_n) = D(\xi_1) + ... + D(\xi_n)$

$D(\hat{L_k}) = \sum \limits_{j=1}^k c^2_j D(\overline X_j) = \sum \limits_{j=1}^k c^2_j \frac{\sigma^2}{n_j} = \sum \limits_{j=1}^k \frac{c^2_j}{n_j} \sigma^2$

$\sigma^2$ дисперсия нам неизвестна, в качестве оценки дисперсии возьмем оценку

$\hat{\sigma^2} = \frac{Q_1}{N-k}$ является оценкой дисперсии в независимости от того, выполняется $H_0$ или нет.

тогда

$S^2_{\hat{L_k}} = \sum \limits_{j=1}^k \frac{c^2_j}{n_j} \frac{Q_1}{N-k}$

$S_{\hat{L_k}} = \sqrt{ \sum \limits_{j=1}^k \frac{c^2_j}{n_j} \frac{Q_1}{N-k} }$



##### Лемма (Метод Шеффе)

Для любой совокупности векторов $(c_1, ..., c_k): \sum \limits_{j=1}^k c_j = 0$ вероятность не меньше 1 - $\alpha$ одновременного выполнения неравенств

$\bigg | \sum \limits_{j=1}^k c_j (a_j - \overline X_j) \bigg | < S_{\hat{L_k}} \cdot \sqrt{(k - 1)} \cdot u_{1-\alpha, k-1, N-k}$, где

$u_{1-\alpha, k-1, N-k}$ - квантиль Фишера.

Это означает, что вероятность отклонится оценки линейного контраста от истинного значения линейного контраста более чем на $S_{\hat{L_k}} \cdot \sqrt{(k - 1)} \cdot u_{1-\alpha, k-1, N-k}$ будет меньше чем $\alpha$.



##### Проверка гипотезы. Интервал Шеффе

$H_0: L_k = 0$

$H_1: L_k \neq 0$

для проверки гипотезы можем воспользоваться интервалом из метода Шеффе

$Interval = (\hat{L_k} - S_{\hat{L_k}} \cdot \sqrt{(k - 1) \cdot u_{1-\alpha, k-1, N-k}} , \;\; \hat{L_k} + S_{\hat{L_k}} \cdot \sqrt{(k - 1) \cdot u_{1-\alpha, k-1, N-k}} )$

Мы принимаем $H_0$ на уровне значимости $\alpha$, если

$0 \in$ $Interval$ $\Rightarrow H_0$ **не отвергается**.

##### Пример

Предположим, что мы сравнивали равенство мат. ож. трех генеральных совокупностей и проверяли гипотезу $H_0: a_1 = a_2 = a_3$. Мы воспользовались однофакторным ДА, и отклонили $H_0$.

Теперь мы хотим выяснить, мат. ож. каких именно генеральных совокупностей не равны, для этого необходимо проверить следующие гипотезы

$H^{12}_0: a_1 = a_2$

$H^{13}_0: a_1 = a_3$

$H^{23}_0: a_2 = a_3$

Воспользуемся методом линейных контрастов, для этого необходимо построить линейный контраст для каждой из гипотез

$H^{12}_0: a_1 = a_2$ $\;$ $L_{k12} = a_1 - a_2$

$H^{13}_0: a_1 = a_3$ $\;$ $L_{k13} = a_1 - a_3$

$H^{23}_0: a_2 = a_3$ $\;$ $L_{k23} = a_2 - a_3$

т.е. $c_1 = 1, c_2 = -1, c_3 = -1$



Далее необходимо построить точечные оценки линейных контрастов 

$\hat{L_{k12}}, \hat{L_{k13}}, \hat{L_{k24}}$ например $\hat{L_{k12}} = \overline X_1 - \overline X_2$

Найти оценки дисперсии оценки линейного контраста

$S^2_{\hat{L_{k12}}}, S^2_{\hat{L_{k13}}}, S^2_{\hat{L_{k23}}}$, например $S^2_{\hat{L_{k13}}} = \frac{1}{n_1} \frac{Q_1}{N-k} + \frac{1}{n_3} \frac{Q_1}{N-k}$, где

$n_1$ - кол-во наблюдений в выборке.

$k$ - кол-во выборок

Остается построить интервалы, согласно методу Шеффе, и проверить, принадлежат ли нули данным интервалам. 

#### Задача №1

Время химической реакции при различном содержании катализатора распределилось следующим образом

5%

5.9, 6.0, 7.0, 6.5, 5.5, 7.0, 8.1, 7.5, 6.2, 6.4, 7.1, 6.9

10% 

4.0, 5.1, 6.2, 5.3, 4.5, 4.4, 5.3, 5.4,5.6,5.2

15% 

8.2, 6.8, 8.0, 7.5, 7.2, 7.9, 8.1, 8.5, 7.8, 8.1

Выборки получены из нормально распределенных генеральных совокупностей с равными дисперсиями. Проверить гипотезу о равенстве средних. Уровень значимости принять равным 0.1.

*  Найдите значение статистики, которая является точечной оценкой дисперсии независимо от того, верна нулевая гипотеза или нет.

*  Определите распределение статистики критерия в предположении справедливости нулевой гипотезы. Найдите квантиль распределения статистики критерия, участвующий в построении критической области. 

*  Найдите выборочное значение статистики критерия и примите статистическое решение.

*  Найдите точечную оценку линейного контраста для проверки частной нулевой гипотезы о равенстве среднего времени химической реакции при содержании катализатора  5% и 10% $(H_0:a_{5\%}=a_{10\%})$ 

*  Найдите доверительный интервал для линейного контраста для проверки частной нулевой гипотезы о равенстве среднего времени химической реакции при содержании катализатора  5% и 10% $(H_0:a_{5\%}=a_{10\%})$ и примите статистическое решение о равенстве сравниваемых мат.ожиданий

Найдите значение статистики, которая является точечной оценкой дисперсии независимо от того, верна нулевая гипотеза или нет.

In [ ]:
X1 = np.array([5.9, 6.0, 7.0, 6.5, 5.5, 7.0, 8.1, 7.5, 6.2, 6.4, 7.1, 6.9])
X2 = np.array([4.0, 5.1, 6.2, 5.3, 4.5, 4.4, 5.3, 5.4,5.6,5.2])
X3 = np.array([8.2, 6.8, 8.0, 7.5, 7.2, 7.9, 8.1, 8.5, 7.8, 8.1])
k = 3
X = np.concatenate((X1, X2, X3), axis=0)
#st.f_oneway(X1, X2, X3)

In [ ]:
n1, n2, n3 = len(X1), len(X2), len(X3)
N = n1 + n2 + n3
xm1, xm2, xm3, xm = X1.mean(), X2.mean(), X3.mean(), X.mean()
Q = sum((X1 - xm1)**2) + sum((X2 - xm2)**2) + sum((X3 - xm3)**2)
statistics = Q / (N - k)
statistics

0.41212068965517235

Определите распределение статистики критерия в предположении справедливости нулевой гипотезы. Найдите квантиль распределения статистики критерия, участвующий в построении критической области. 

$V_k = (u_{1 - \alpha, k - 1, N - k}; +\infty)$, где

$u_{1 - \alpha, k - 1, N - k}$ - квантиль Фишера.

In [ ]:
alpha = 0.1
dfn, dfd = k - 1, N - k
u = st.f.ppf(1 - alpha, dfn, dfd)
u

2.4954833142354644

Найдите выборочное значение статистики критерия и примите статистическое решение

$F = \frac{Q_2 / (k-1)}{Q_1 / (N-k)}$

In [ ]:
Q1 = sum((X1 - xm1)**2) + sum((X2 - xm2)**2) + sum((X3 - xm3)**2)
Q2 = n1 * (xm1 - xm)**2 + n2 * (xm2 - xm)**2 + n3 * (xm3 - xm)**2 

numerator = Q2 / (k-1)
denominator = Q1 / (N-k)
Fv = numerator / denominator
Fv

44.991068066769834

Найдём квантиль распределения Фишера для построения критической области.

$u_{1 - \alpha, k - 1, N - k}$

In [ ]:
u = st.f.ppf(1 - alpha, k-1, N-k)
u

2.4954833142354644

$V_k = (2.49, + \infty)$

$F_в \in V_k \Rightarrow H_0$ отвергается.

Найдите точечную оценку линейного контраста для проверки частной нулевой гипотезы о равенстве среднего времени химической реакции при содержании катализатора  5% и 10% $(H_0:a_{5\%}=a_{10\%})$

Будем проверять гипотезу

$H^{5\%, 10\%}_0: a_{5\%} = a_{10\%}$ $\;$ $L_{k12} = a_{5\%} - a_{10\%}$

Необходимо найти 

$\hat{L_{k5\%, 10\%}} = \overline X_1 - \overline X_2$

In [ ]:
L5_10 = xm1 - xm2
L5_10

1.5750000000000002

Примите статистическое решение о равенстве сравниваемых мат.ожиданий

Необходимо найти интервал по методу Шеффе

$Interval = (\hat{L_k} - S_{\hat{L_k}} \cdot \sqrt{(k - 1) \cdot u_{1-\alpha, k-1, N-k}} , \;\; \hat{L_k} + S_{\hat{L_k}} \cdot \sqrt{(k - 1) \cdot u_{1-\alpha, k-1, N-k} } )$

Так же необходимо найти точечную оценку среднеквадратичного отклонения (корень из дисперсии)

$S^2_{\hat{L_{k5\%10\%}}} = \frac{1}{n_{5\%}} \frac{Q_1}{N-k} + \frac{1}{n_{10\%}} \frac{Q_1}{N-k}$

$S_{\hat{L_{k5\%10\%}}} = \sqrt{ \frac{1}{n_{5\%}} \frac{Q_1}{N-k} + \frac{1}{n_{10\%}} \frac{Q_1}{N-k} }$

$Q_1 = \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X_j)^2$

$Q_2 = \sum \limits_{j=1}^k n_j (\overline X_j - \overline X)^2$

In [ ]:
S = np.sqrt( 1/n1 * Q1 / (N-k) + 1/n2 * Q1 / (N-k) )
S

0.27487353413909266

Найдём концы интервала

In [ ]:
left = L5_10 - (S * np.sqrt((k-1) * u))
right = L5_10 + (S * np.sqrt((k-1) * u))

left, right

(0.9609195669039591, 2.1890804330960414)

Чтобы принять или отвергнуть гипотезу $H^{5\%, 10\%}_0: a_{5\%}$ нужно проверить, принадлежит или нет ноль этому интервалу.

$0 \notin (0.6, 2.54) \Rightarrow H^{5\%, 10\%}_0$ отвергается.

Значит мат. ож. $a_{5\%} = a_{10\%}$ равны.

**Ответ:**

$\frac{Q_1}{N-k} = 0.412$

Квантиль распределения Фишера участвующий в построении критической области $V_k$

$u_{1 - \alpha, k - 1, N - k} = 2.49$

$H_0$ отвергается. Значит различная концентрация катализатора влияет на время химической реакции.

$\hat{L_{k5\%, 10\%}} = \overline X_1 - \overline X_2 = 1.575$

Интервал Шеффе $(0.96, 2.189)$

$H^{5\%, 10\%}_0$ отвергается

#### Задача №2

Cравните среднюю продолжительность успешных проходов в 10 матчах по рэгби.

Уровень значимости 0.05. Предполагаем, что продолжительность успешных проходов имеет нормальное распределение.

Чему равна квантиль распределения статистики критерия, входящий в определение критической области?

In [ ]:
df = pd.read_csv('rugby.txt', sep='\t')
df.head(2)

Game  Time
0     1  39.2
1     1   2.7

**Решение:**

Необходимо проверить гипотезу

$H_0: a_1 = a_2 = ... = a_k$

$H_1: \exists t,s: a_t \neq a_s$

Воспользуемся параметрическим критерием ДА.

$F = \frac{Q_2 / (k-1)}{Q_1 / (N-k)} \sim F_{(k-1, N-k)}$, где

$F_{k-1, N-k}$ - ф-я распределения Фишера.

$Q_1 = \sum \limits_{j=1}^k \sum \limits_{i=1}^{n_j} (x_{i,j} - \overline X_j)^2$

$Q_2 = \sum \limits_{j=1}^k n_j (\overline X_j - \overline X)^2$

И составим критическую область

$V_k = (u_{1 - \alpha, k - 1, N - k}; +\infty)$, где

$u_{1 - \alpha, k - 1, N - k}$ - квантиль Фишера.

In [ ]:
def grouping(df):
  Xmeans = {}
  num_groups = df['Game'].unique()

  for i in num_groups:
    group = df[df['Game'] == i]['Time'].values
    Xmeans.update({f'{i}': group.mean()})

  return Xmeans

Реализуем вычисление статистического критерия $F$ через $scipy$ библиотеку.

In [ ]:
def grouping(df):
  groups = []
  num_groups = df['Game'].unique()

  for i in num_groups:
    group = df[df['Game'] == i]['Time'].values
    groups.append(group)

  return groups

In [ ]:
groups = grouping(df)

In [ ]:
# F = 3.886
st.f_oneway(*groups)

F_onewayResult(statistic=3.8867057658980024, pvalue=7.334946486662759e-05)

Реализуем вычисление статистического критерия $F$ используя numpy, pandas и чистый python

In [ ]:
import pandas as pd
import numpy as np
from scipy.stats.distributions import f

# prepare data
a = pd.read_csv('rugby.txt', sep='\t')
a = a.groupby('Game')['Time'].apply(lambda df: df.reset_index(drop=True)).unstack()
a = [x[np.logical_not(np.isnan(x))] for x in a.values]
# по итогу получаем список np.ndarray объектов

alpha = 0.05
N = sum(len(x) for x in a)
k = len(a)
m = np.concatenate(a).mean()
f_local = f(k - 1, N - k)

Q1 = sum(sum((x - x.mean())**2) for x in a)
Q2 = sum(len(x) * (m - x.mean())**2 for x in a)

Fv = (Q2 / (k - 1)) / (Q1 / (N - k))
print(f"Выборочное значение статистики: {Fv}")
print(f"Критическая область: ({f_local.ppf(1-alpha)}; +inf)")
print(f"p-value: {1 - f_local.cdf(Fv)}")

Выборочное значение статистики: 3.886705765898003
Критическая область: (1.889525776660891; +inf)
p-value: 7.334946486659e-05


Реализация вычисления F

In [ ]:
def get_Q1(df):
  num_groups = df['Game'].unique()
  Q1 = .0

  for i in num_groups:
    group = df[df['Game'] == i]['Time'].values
    Q1 += sum((group - group.mean())**2)
  
  return Q1

In [ ]:
def get_Q2(df):
  Xmean = df['Time'].values.mean()
  num_groups = df['Game'].unique()
  Q2 = .0

  for i in num_groups:
    group = df[df['Game'] == i]['Time'].values
    n = len(group)
    Q2 += n * (group.mean() - Xmean)**2
  
  return Q2

In [ ]:
def get_F(Q1, Q2, df):
  k = df['Game'].nunique()
  N = df['Game'].count()

  F = (Q2 / (k-1)) / (Q1 / (N-k))

  return F

In [ ]:
def get_quantile(df, alpha = 0.05):
  k = df['Game'].nunique()
  N = df['Game'].count()

  u = st.f.ppf(1 - alpha, k - 1, N - k)

  return u

In [ ]:
Q1 = get_Q1(df)
Q2 = get_Q2(df)
F = get_F(Q1, Q2, df)
u = get_quantile(df, alpha = 0.05)

print(f'F: {F:.3f}')
print('H0 отвергается' if F > u else 'H0 не отвергается')

F: 3.887
H0 отвергается


**Ответ:**

$F_в = 3.887$

$H_0$ отвергается


#### Задача №3

У студентов измерили пульс до и после тренировки. Верно ли, что пульс в среднем увеличился на 50 ударов в минуту?

Проверьте данное предположение с помощью доверительных интервалов. Уровень значимости 0.02

Постройте приближенный доверительный интервал для $a_1-a_2$. Примите статистическое решение.

##### Решение

##### Гипотеза

Нам необходимо проверить гипотезу построив ассимптотический дов. интервал

$H_0: a_1 = a_2 - 50$

$H_1: a_1 \neq a_2 - 50$

##### Построение доверительного интервала

Предположим, что наши генеральные совокупности имеют нормальный закон распределения $N(a, \sigma)$

Мы знаем, что 

$\sqrt{n} (\hat{\theta} - \theta) \xrightarrow[n \rightarrow +\infty]{d_\theta} \sim N(0, \sigma)$

$P(u_{\frac{\alpha}{2}} < \frac{\sqrt{n} (\hat{\theta} - \theta)}{\hat{\sigma}} < u_{1 - \frac{\alpha}{2}} ) = 1 - \alpha$

$P(\hat{\theta} \pm u_{1 - \frac{\alpha}{2}} \frac{\hat{\sigma}}{\sqrt{n}} ) = 1 - \alpha$

Тогда интервал будет равен

$ \bigg ( \hat{\theta} \pm u_{1 - \frac{\alpha}{2}} \frac{\hat{\sigma}}{\sqrt{n}} \bigg )$, где

$u_{1 - \frac{\alpha}{2}}$ квантиль стандартного нормального распределения

##### Построение точечных оценок

Обозначим генеральную совокупность измерения пульса до тренировки $\xi_1$ после тренировки $\xi_2$

Построим сл. вел. равную $Y = \xi_1 - \xi_2$

т.к. мы предполагаем, что ген. сов. имеет нормальный закон распр., то и точечные оценки выброчного среднего так же имеют норм. зак. распр., тогда и 

$Y = \xi_1 - \xi_2 \sim N(a, \sigma)$

$a_1$ мат. ож. ген. сов. $\xi_1$

$a_2$ мат. ож. ген. сов. $\xi_2$

В качестве параметра возьмем $\theta = a_1 - a_2$

В качестве точечных оценок мат. ож. возьмем $a_1 = \overline X_1$ и $a_2 = \overline X_2$

Тогда оценка будет равна

$\hat{\theta} = \overline X = \overline X_1 - \overline X_2$

В качестве точечных оценок дисперсии, возьмем несмещенную оценку выборочной дисперсии

$\hat{\sigma^2} = \frac{1}{n-1} \sum\limits_{i=1}^n (y_i - \overline X)^2$

$ \hat{\sigma} = \sqrt{ \frac{1}{n-1} \sum\limits_{i=1}^n (y_i - \overline X)^2 } $, где

$y_i = x_{1i} - x_{2i}$, где $x_{1i} \in \xi_1$, $x_{2i} \in \xi_2$, $i$-ое наблюдение, $i = \overline {1,n}$

##### Построение статистического критерия

т.к. наши случайные величины **имеют нормальный закон распределения**, а **выборки зависимы** (т.к. измерения производятся на одном и том же объекте), в качестве статистического критерия возьмем **параметрический критерий Стьюдента**, при большом количестве наблюдений, оно стремится к нормальму распределению.

В кач-ве статистики критерия возьмем

$Z = \frac{\overline X}{\hat{\sigma} / \sqrt{n}} \sim T_{n-1}$, где

$T_{n-1}$ - распределение Стьюдента

In [ ]:
df = pd.read_csv('pulse.txt', sep = '\t')
df.head(2)

Pulse1  Pulse2
0      82     150
1      96     176

Найдём $\overline X = \overline X_1 - \overline X_2$

In [ ]:
Xmean = df['Pulse1'].mean() - df['Pulse2'].mean()
Xmean

-51.39130434782608

Найдём $ \hat{\sigma} = \sqrt{ \frac{1}{n-1} \sum\limits_{i=1}^n (y_i - \overline X)^2 } $

In [ ]:
n = df['Pulse1'].count()
x1 = df['Pulse1'].values
x2 = df['Pulse2'].values
Y = x1 - x2
S = np.sqrt( sum((Y - Xmean)**2) / (n-1) )
S

21.089732373697938

Найдём $u_{1 - \frac{\alpha}{2}}$

In [ ]:
u = st.norm.ppf(1 - (0.02/2))
u

2.3263478740408408

Найдём интервал $ \bigg ( \hat{\theta} \pm u_{1 - \frac{\alpha}{2}} \frac{\hat{\sigma}}{\sqrt{n}} \bigg )$

In [ ]:
left = Xmean - (u * (S / np.sqrt(n)))
right = Xmean + (u * (S / np.sqrt(n)))

left, right

(-58.62510957312638, -44.15749912252578)

##### Вычислим интервал используя Scipy

In [ ]:
a1, a2 = df['Pulse1'], df['Pulse2']
a = a1 - a2
alpha = 0.02

# st.sem - стандартная ошибка среднего
st.norm.interval(1 - alpha, loc=np.mean(a), scale=st.sem(a))

(-58.62510957312639, -44.157499122525785)

$V_k = (-58.62, -44.15)$

Вычислим статистический критерий 

$Z = \frac{\overline X}{\hat{\sigma} / \sqrt{n}}$

In [ ]:
zv = np.sqrt(n) * Xmean / S
zv

-16.52713169489407

$Z_в \notin V_k \Rightarrow H_0$ не отвергается. Значит пульс в среднем увеличился, после тренировки.

**Ответ:**

$V_k = (-58.62, -44.15)$

$H_0$ не отвергается

#### Непараметрический критерий. Однофакторный ДА

##### Критерий Краскела-Уоллиса

Мы исследуем одномерную зависимую переменную, на которую влияет всего 1 фактор, принимающий $k>2$ значений.

Выборки случайны и взаимно независимы

Генеральные совокупности имеют непрервные ф-и распределения $F_1, ..., F_k$

$X_{11} \;\; X_{12} \;\; ... \;\; X_{1k}$

$X_{21} \;\; X_{22} \;\; ... \;\; X_{2k}$

$\;\; . \;\;\;\;\;\;\;\; . \;\;\;\; ... \;\;\;\; . \;\;$

$\;\; . \;\;\;\;\;\;\;\; . \;\;\;\; ... \;\;\;\; . \;\;$

$X_{n_11} \;\; X_{n_22} \;\; ... \;\; X_{n_kk}$

$\;\;\;\; \xi_1 \;\;\;\;\;\; \xi_2 \;\;\; ... \;\;\;\; \xi_k$

$x_{i, j}$ $i$-ый элемент, $j$-ой выборки

##### Проверяем гипотезу

$H_0: F_1(x) = F_2(x) = ... = F_k(x)$ для всех $x \in \mathbb R$

$H_1: F_1(x) = F_2(x - \Delta_2) = ... = F_k(x - \Delta_k)$ для всех $x \in \mathbb R$

если отвергается $H_0$ значит фактор оказывает влияния на исследуемую переменную.

Критерий Краскела-Уоллиса является обобщением критерия Вилкоксона на случай сравнения распределения нескольких генеральных совокупностей.

Объединим выборки в одну большу, кол-во наблюдей в ней будет

$N = \sum \limits_{j=1}^k n_j$

Составим вариационный ряд

$x_{(1)} \leq x_{(2)} \leq ... \leq x_{(N)}$

Обозначим ранг

$R_{i,j} = rank(x_{i,j})$, где

x_{i,j} - $i$-ый элемент, $j$-ой выборки.


##### Статистика критерия H

Средний ранг $j$-ой выборки

$ \overline R_j = \frac{1}{n_j} \sum \limits_{i=1}^{n_j} R_{i,j}$

Общий средний ранг

$ \overline R = \frac{1}{N} \sum\limits_{i=1}^k \sum\limits_{j=1}^{n_j} R_{i,j} = \frac{N+1}{2}$

Статистика критерия

$Q_2 = \sum\limits_{i=1}^k n_j (\overline R_j - \overline R)^2$

если $H_0$ отвергается, то $Q_2$ принимает значения гораздо большие, чем когда $H_0$ не отвергается.

Можно было бы строить критическую область $V_k$ прямо следуя подходу принятому для критерия Вилкоксона, но на практике используют статистику $H$

$H = \frac{12}{N(N+1)} Q_2 = \frac{12}{N(N+1)} \sum\limits_{j=1}^k n_j \bigg ( \overline R_j - \frac{N+1}{2} \bigg )^2 = \\
= \frac{12}{N(N+1)} \bigg( \sum \limits_{j=1}^k n_j \overline R^2_j\bigg ) - 3(N+1) \sim \chi^2_{k-1}$

в случае, когда

$k=3$ и $n_j \geq 5$ 

$k>3$ и $n_j \geq 4$ 

используется приближенное распределение $H \sim \chi^2_{k-1}$. Если данные условия не выполняются, то используют таблицы критических значений статистики $H$.

##### Статистика критерия в Scipy

Функция для вычисления статистики $H$ из библиотеки $scipy.stats$, в качестве аргументов передаём сортированные выборки

$st.kruskal(variation \: range)$

#### Критическая область

$V_k = (\chi^2_{1 - \alpha, k-1}, +\infty)$

$H_в \in V_k \Rightarrow H_0$ отвергается. Считаем, что фактор влияет на зависимую переменную.

##### Модификации

В случае, если в объединенной выборке есть элементы из разных выборок с повторяющимися значениями, тогда необходимо скорректировать статистику $H$ и использовать $H^*$

$H^* = H \bigg (1 - \bigg ( \sum \limits_{j=1}^q \frac{t^3_j - t_j}{N^3 - N} \bigg ) \bigg )^{-1}$, где

$q$ кол-во связок, кол-во последовательностей с одинаковыми значениями из разных выборок.

$t_j$ размер $j$-ой связки, то кол-во эл-ов, которые входят в $j$-ю связку.

Для того, чтобы вычислить ранги в случае с повторяющимися значениями, нам необходимо взять среднее арифмитическое рангов как если бы эти элементы имели разные значения.

**Пример**

1, 3, 5, 3 выборка 

1, 3, 3, 5 вариационный ряд 

$rank(1) = 1, rank(3) = 2,5, rank(5) = 4, rank(3) = 2,5$


##### Статистика Имана-Давенпорта

При больших $N$ использют аппроксимацию Имана-Давенпорта. Используются статистику $J$

$J = \frac{H}{2} \bigg ( 1 + \frac{N-k}{N-1-H}  \bigg )$

##### Критическая область

Все значения J, которые превосходят сумму квантилей Фишера и хи-квадрат

$V_k = \bigg ( J > \bigg ( F_{1 - \alpha, k-1, N-k} \cdot (k-1) + \chi^2_{1 - \alpha, k-1}  \bigg ) \bigg )$

$J_в \in V_k \Rightarrow H_0$ отклоняем. Значит фактор влияет на зависимую переменную.

##### Задача №1

Проверьте, правда ли, что болевой барьер у блондинов и брюнетов разный.

Светлый блондин: 62, 60, 71, 55, 48

Темный блондин: 63, 57, 52, 41, 43

Светлый брюнет: 42, 50, 44, 37

Темный брюнет: 32, 39, 51, 30, 35

*  Найдите средний ранг болевого барьера для  светлых брюнетов.

*  Найдите выборочное значение статистики критерия и примите статистическое решение. Уровень значимости 0.05

**Решение:**

Найдите средний ранг болевого барьера для  светлых брюнетов.

Нам необходимо найти

$ \overline R_j = \frac{1}{n_j} \sum \limits_{i=1}^{n_j} R_{i,j}$

Вначале необходимо составить одну большую выборку, из неё сделать вариационный ряд. По вариационному рядку искать ранг из подвыборки светлых брюнетов.

In [ ]:
sbl = np.array([62, 60, 71, 55, 48])
tbl = np.array([63, 57, 52, 41, 43])
sbr = np.array([42, 50, 44, 37])
tbr = np.array([32, 39, 51, 30, 35])

X = np.concatenate((sbl, tbl, sbr, tbr))
variational_range = np.sort(X)
# + 1 т.к. индексация идёт с 0, нам нужно с 1
rank = np.searchsorted(variational_range, sbr) + 1
R_sbr = rank.mean()
R_sbr

7.75

Найдите выборочное значение статистики критерия и примите статистическое решение. Уровень значимости 0.05

Нам необходимо найти выборочную статистикукритерия $H$.

$H = \frac{12}{N(N+1)} Q_2 = \frac{12}{N(N+1)} \sum\limits_{j=1}^k n_j \bigg ( \overline R_j - \frac{N+1}{2} \bigg )^2 = \sim \chi^2_{k-1}$

в случае, когда

$k=3$ и $n_j \geq 5$ 

$k>3$ и $n_j \geq 4$ 

и построить критическую область 

$V_k = (\chi^2_{1 - \alpha, k-1}, +\infty)$

In [ ]:
N = len(X)
n1, n2, n3, n4 = len(sbl), len(tbl), len(sbr), len(tbr)
R_sbl = (np.searchsorted(variational_range, sbl) + 1).mean()
R_tbl = (np.searchsorted(variational_range, tbl) + 1).mean()
R_sbr = (np.searchsorted(variational_range, sbr) + 1).mean()
R_tbr = (np.searchsorted(variational_range, tbr) + 1).mean()

Q = n1 * (R_sbl - (N+1)/2)**2 + n2 * (R_tbl - (N+1)/2)**2 + n3 * (R_sbr - (N+1)/2)**2 + n4 * (R_tbr - (N+1)/2)**2

Hv = (12 / (N * (N+1))) * Q
Hv

10.144736842105264

Функция для вычисления статистики $H$ из библиотеки $scipy.stats$

$st.kruskal(variation \; range)$

In [ ]:
st.kruskal(np.sort(sbl), np.sort(tbl), np.sort(sbr), np.sort(tbr))

10.14473684210526

In [ ]:
k = 4
alpha = 0.05
st.chi2.ppf(1 - alpha, k - 1)

7.814727903251179

$V_k = (7.8, + \infty)$

$Hv \in V_k \Rightarrow H_0$ отвергается.


**Ответ:**

Средний ранг болевого барьера для  светлых брюнетов

$\overline R_{sbr} = 7.75$

$H_в = 10.145$

$H_0$ отвергается